File: day07/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	var nums []int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		for _, i := range strings.Split(scanner.Text(), ",") {
			num, err := strconv.Atoi(i)
			if err != nil {
				log.Fatal(err)
			}
			nums = append(nums, num)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	sort.Ints(nums)
	median := nums[len(nums)/2]
	sum := 0
	for _, i := range nums {
		sum += i
	}
	a1 := int(math.Floor(float64(sum) / float64(len(nums))))
	a2 := int(math.Ceil(float64(sum) / float64(len(nums))))
	diffs := 0
	diffs1 := 0
	diffs2 := 0
	for _, i := range nums {
		diffs += abs(i, median)
		k1 := abs(i, a1)
		diffs1 += k1 * (k1 + 1) / 2
		k2 := abs(i, a2)
		diffs2 += k2 * (k2 + 1) / 2
	}
	fmt.Println("Part 1:", diffs)
	fmt.Println("Part 2:", min(diffs1, diffs2))
}

func abs(a, b int) int {
	if a < b {
		return b - a
	}
	return a - b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: day07/src/main.rs
use std::cmp::min;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut nums = Vec::new();
    for line in reader.lines() {
        nums = line?
            .split(",")
            .flat_map(|i| i.parse::<i32>())
            .collect::<Vec<_>>();
    }
    nums.sort();
    let median = nums[nums.len() / 2];
    let (a1, a2) = {
        let k = nums.iter().sum::<i32>();
        let a = (k as f64) / nums.len() as f64;
        (a.floor() as i32, a.ceil() as i32)
    };
    let (d, d1, d2) = nums.iter().fold((0, 0, 0), |(d, d1, d2), i| {
        let k1 = (i - a1).abs();
        let k2 = (i - a2).abs();
        (
            d + (i - median).abs(),
            d1 + k1 * (k1 + 1) / 2,
            d2 + k2 * (k2 + 1) / 2,
        )
    });
    println!("Part 1: {}", d);
    println!("Part 2: {}", min(d1, d2));
    Ok(())
}
